In [ ]:
import pandas as pd
df_ol_sales = pd.read_csv('/content/online_sales_dataset.csv')

df_ol_sales

InvoiceNo StockCode   Description  Quantity       InvoiceDate  \
0         221958  SKU_1964     White Mug        38  2020-01-01 00:00   
1         771155  SKU_1241     White Mug        18  2020-01-01 01:00   
2         231932  SKU_1501    Headphones        49  2020-01-01 02:00   
3         465838  SKU_1760     Desk Lamp        14  2020-01-01 03:00   
4         359178  SKU_1386     USB Cable       -30  2020-01-01 04:00   
...          ...       ...           ...       ...               ...   
49777     354083  SKU_1562      Blue Pen        25  2025-09-05 01:00   
49778     296698  SKU_1930     USB Cable         7  2025-09-05 02:00   
49779     177622  SKU_1766  Office Chair        43  2025-09-05 03:00   
49780     701213  SKU_1602      Notebook        48  2025-09-05 04:00   
49781     772215  SKU_1832     White Mug        30  2025-09-05 05:00   

       UnitPrice  CustomerID         Country  Discount  PaymentMethod  \
0           1.71     37039.0       Australia  0.470000  Bank Transfer   
1          41.25     19144.0           Spain  0.190000        paypall   
2          29.11     50472.0         Germany  0.350000  Bank Transfer   
3          76.68     96586.0     Netherlands  0.140000        paypall   
4         -68.11         NaN  United Kingdom  1.501433  Bank Transfer   
...          ...         ...             ...       ...            ...   
49777      70.92     51445.0           Spain  0.200000    Credit Card   
49778      51.74     28879.0   United States  0.230000  Bank Transfer   
49779      85.25     21825.0        Portugal  0.200000  Bank Transfer   
49780      39.64     43199.0  United Kingdom  0.310000        paypall   
49781      38.27     53328.0          France  0.100000    Credit Card   

       ShippingCost     Category SalesChannel  ReturnStatus ShipmentProvider  \
0             10.79      Apparel     In-store  Not Returned              UPS   
1              9.51  Electronics       Online  Not Returned              UPS   
2             23.03  Electronics       Online      Returned              UPS   
3             11.08  Accessories       Online  Not Returned       Royal Mail   
4               NaN  Electronics     In-store  Not Returned            FedEx   
...             ...          ...          ...           ...              ...   
49777          8.96  Electronics       Online      Returned              UPS   
49778         23.55  Electronics       Online  Not Returned            FedEx   
49779         16.26    Furniture     In-store  Not Returned            FedEx   
49780         28.56      Apparel       Online  Not Returned       Royal Mail   
49781          9.13   Stationery       Online  Not Returned              UPS   

      WarehouseLocation OrderPriority  
0                London        Medium  
1                  Rome        Medium  
2                Berlin          High  
3                  Rome           Low  
4                   NaN        Medium  
...                 ...           ...  
49777            Berlin        Medium  
49778         Amsterdam           Low  
49779            London          High  
49780            London        Medium  
49781              Rome           Low  

[49782 rows x 17 columns]

In [ ]:
df_ol_sales.describe()
df_ol_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49782 entries, 0 to 49781
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   InvoiceNo          49782 non-null  int64  
 1   StockCode          49782 non-null  object 
 2   Description        49782 non-null  object 
 3   Quantity           49782 non-null  int64  
 4   InvoiceDate        49782 non-null  object 
 5   UnitPrice          49782 non-null  float64
 6   CustomerID         44804 non-null  float64
 7   Country            49782 non-null  object 
 8   Discount           49782 non-null  float64
 9   PaymentMethod      49782 non-null  object 
 10  ShippingCost       47293 non-null  float64
 11  Category           49782 non-null  object 
 12  SalesChannel       49782 non-null  object 
 13  ReturnStatus       49782 non-null  object 
 14  ShipmentProvider   49782 non-null  object 
 15  WarehouseLocation  46297 non-null  object 
 16  OrderPriority      497

In [ ]:
df_ol_sales['Country'].value_counts()

Country
France            4230
Sweden            4211
Germany           4182
United Kingdom    4180
Netherlands       4173
Belgium           4170
Portugal          4163
Norway            4157
Australia         4110
Spain             4100
United States     4058
Italy             4048
Name: count, dtype: int64

In [ ]:
df_ol_sales['Description'].value_counts()

Description
Wall Clock        4617
USB Cable         4580
Headphones        4555
Backpack          4550
Desk Lamp         4545
White Mug         4536
Office Chair      4522
Blue Pen          4509
T-shirt           4466
Notebook          4454
Wireless Mouse    4448
Name: count, dtype: int64

In [ ]:
df_ol_sales['Category'].value_counts()

Category
Furniture      10084
Accessories     9975
Electronics     9931
Stationery      9928
Apparel         9864
Name: count, dtype: int64

In [ ]:
#Sales_Analysis
df_ol_sales['TotalSales'] = df_ol_sales['UnitPrice'] * df_ol_sales['Quantity'] * (1 - df_ol_sales['Discount'])
category_sales = df_ol_sales.groupby('Category')['TotalSales'].sum()
print(category_sales)

Category
Accessories    8.731408e+06
Apparel        8.684128e+06
Electronics    8.636717e+06
Furniture      8.801411e+06
Stationery     8.657717e+06
Name: TotalSales, dtype: float64


In [ ]:
#Customer_Analysis
customer_sales = df_ol_sales.groupby('CustomerID')['TotalSales'].sum()
print(customer_sales)

CustomerID
10001.0    1470.7334
10003.0     365.7385
10005.0    2350.1430
10008.0      48.8280
10009.0     463.5904
             ...    
99986.0    1975.0840
99989.0      24.2022
99993.0     299.3744
99997.0    1563.5640
99998.0    2638.5314
Name: TotalSales, Length: 35389, dtype: float64


In [ ]:
#Time_of_Sale
df_ol_sales['InvoiceDate'] = pd.to_datetime(df_ol_sales['InvoiceDate'])
monthly_sales = df_ol_sales.resample('M', on='InvoiceDate')['TotalSales'].sum()
print(monthly_sales)

InvoiceDate
2020-01-31    630703.010252
2020-02-29    608863.278708
2020-03-31    648736.022424
2020-04-30    660908.151737
2020-05-31    644803.553348
                  ...      
2025-05-31    651934.977776
2025-06-30    597415.753604
2025-07-31    635611.162991
2025-08-31    661838.894834
2025-09-30     87494.000717
Freq: ME, Name: TotalSales, Length: 69, dtype: float64


<ipython-input-12-3bfcba55a383>:3: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  monthly_sales = df_ol_sales.resample('M', on='InvoiceDate')['TotalSales'].sum()


In [ ]:
#Geographical_Analysis
country_sales = df_ol_sales.groupby('Country')['TotalSales'].sum()
print(country_sales)

Country
Australia         3.498378e+06
Belgium           3.752605e+06
France            3.648590e+06
Germany           3.672380e+06
Italy             3.562899e+06
Netherlands       3.470682e+06
Norway            3.577973e+06
Portugal          3.610836e+06
Spain             3.613235e+06
Sweden            3.673576e+06
United Kingdom    3.714717e+06
United States     3.715511e+06
Name: TotalSales, dtype: float64


In [ ]:
#Product_Analysis
top_products = df_ol_sales.groupby('Description')['TotalSales'].sum().sort_values(ascending=False).head(10)
print(top_products)

Description
White Mug         4.086721e+06
USB Cable         4.048808e+06
Desk Lamp         4.001349e+06
Backpack          4.000374e+06
Wall Clock        3.962512e+06
Blue Pen          3.934804e+06
Headphones        3.923602e+06
Wireless Mouse    3.922392e+06
T-shirt           3.898712e+06
Notebook          3.867359e+06
Name: TotalSales, dtype: float64


In [ ]:
#Least_sold_product
least_sold_product = df_ol_sales.groupby('Description')['Quantity'].sum().idxmin()
print(least_sold_product)

Headphones


In [ ]:
#Most_sold_product
most_sold_product = df_ol_sales.groupby('Description')['Quantity'].sum().idxmax()
print(most_sold_product)

USB Cable


In [ ]:
#Return_Analysis
return_rate = df_ol_sales[df_ol_sales['ReturnStatus'] == 'Returned'].shape[0] / df_ol_sales.shape[0]
print(return_rate)

0.09830862560764936


In [ ]:
#Profit
df_ol_sales['Profit'] = df_ol_sales['TotalSales'] - df_ol_sales['ShippingCost']
print(df_ol_sales[['Description', 'Profit']])

        Description     Profit
0         White Mug    23.6494
1         White Mug   591.9150
2        Headphones   904.1235
3         Desk Lamp   912.1472
4         USB Cable        NaN
...             ...        ...
49777      Blue Pen  1409.4400
49778     USB Cable   255.3286
49779  Office Chair  2916.3400
49780      Notebook  1284.3168
49781     White Mug  1024.1600

[49782 rows x 2 columns]


In [ ]:
# Group by 'Description' and aggregate the necessary columns
grouped = df_ol_sales.groupby('Description').agg({ 'TotalSales': 'sum', 'ShippingCost': 'sum' }).reset_index()

# Calculate profit for each product
grouped['Profit'] = grouped['TotalSales'] - grouped['ShippingCost']

# Display the results
print(grouped)

       Description    TotalSales  ShippingCost        Profit
0         Backpack  4.000374e+06      75499.40  3.924875e+06
1         Blue Pen  3.934804e+06      74560.84  3.860243e+06
2        Desk Lamp  4.001349e+06      75769.19  3.925579e+06
3       Headphones  3.923602e+06      75420.18  3.848182e+06
4         Notebook  3.867359e+06      74734.37  3.792624e+06
5     Office Chair  3.864750e+06      74442.13  3.790308e+06
6          T-shirt  3.898712e+06      74949.44  3.823762e+06
7        USB Cable  4.048808e+06      77176.50  3.971631e+06
8       Wall Clock  3.962512e+06      76028.85  3.886483e+06
9        White Mug  4.086721e+06      75129.23  4.011591e+06
10  Wireless Mouse  3.922392e+06      73658.64  3.848733e+06


In [ ]:
#Category Sales Insights
category_sales = df_ol_sales.groupby('Category')['TotalSales'].sum()
print(category_sales)

Category
Accessories    8.731408e+06
Apparel        8.684128e+06
Electronics    8.636717e+06
Furniture      8.801411e+06
Stationery     8.657717e+06
Name: TotalSales, dtype: float64


In [ ]:
#Payment_method_Analysis
payment_method_sales = df_ol_sales.groupby('PaymentMethod')['TotalSales'].sum()
print(payment_method_sales)

PaymentMethod
Bank Transfer    1.476484e+07
Credit Card      1.441177e+07
paypall          1.433477e+07
Name: TotalSales, dtype: float64


Here’s a concise summary of online sales:

Total Sales Calculation:

Each product's total sales were calculated by multiplying unit price by quantity and adjusting for discounts.

Total Sales: [Category
Accessories    8.73M
Apparel        8.68M
Electronics    8.63M
Furniture      8.80M
Stationery     8.65M]

Profit Calculation:

Profit for each product was calculated by subtracting total shipping costs from total sales.


Grouped Data: [Description    TotalSales  ShippingCost        Profit
0         Backpack               4.00M      75499.40           3.92M
1         Blue Pen  3.93M      74560.84  3.86M
2        Desk Lamp  4.00M      75769.19  3.92M
3       Headphones  3.92M      75420.18  3.84M
4         Notebook  3.86M      74734.37  3.79M
5     Office Chair  3.86M      74442.13  3.79M
6          T-shirt  3.89M      74949.44  3.82M
7        USB Cable  4.04M      77176.50  3.97M
8       Wall Clock  3.96M      76028.85  3.88M
9        White Mug  4.08M      75129.23  4.01M
10  Wireless Mouse  3.92M      73658.64  3.84M]


Key Findings:

Most Sold Product: [USB Cable]

Least Sold Product: [Headphones]


Category Sales Insights:

Sales distribution across different product categories was analyzed.

Top Categories: [Category
Accessories    8.731408e+06
Apparel        8.684128e+06
Electronics    8.636717e+06
Furniture      8.801411e+06
Stationery     8.657717e+06]

Geographical Analysis:

Sales data was analyzed geographically to identify distribution across different countries.

Country Sales: [Country
Australia         3.498378e+06
Belgium           3.752605e+06
France            3.648590e+06
Germany           3.672380e+06
Italy             3.562899e+06
Netherlands       3.470682e+06
Norway            3.577973e+06
Portugal          3.610836e+06
Spain             3.613235e+06
Sweden            3.673576e+06
United Kingdom    3.714717e+06
United States     3.715511e+06]